File: server/src/rest_api/schemas/players/players.py
from datetime import date

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

from common.types import PlayerPosition


class Player(BaseModel):
    """
    プレイヤーの基本情報を示します.
    """

    player_id: int
    """プレイヤーID"""
    team_id: int | None
    """所属チームID"""
    full_name: str
    """名称"""
    abbreviation: str
    """略称"""
    position: PlayerPosition | None
    """ポジション"""
    date_of_birth: date | None
    """生年月日"""
    draft_year: int | None
    """ドラフト年"""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )


class GamePlayer(Player):
    jearsy_num: str
    """背番号"""
    is_home: bool
    """ホームチームか否か"""
    is_starter: bool
    """スターティングメンバーか否か"""
    is_active: bool
    """アクティブか否か"""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )

File: client/app/api/schemas/players.ts
/**
 * プレイヤーのポジションを定義します.
 */
export const IPosition = {
  PointGuard: "PG",
  ShootingGuard: "SG",
  SmallForward: "SF",
  PowerForward: "PF",
  Center: "C",
} as const;

/**
 * プレイヤーのポジションを定義します.
 */
export type IPosition = (typeof IPosition)[keyof typeof IPosition];

/**
 * API から受け取る選手情報を表す schema です.
 */
interface IPlayer {
  playerId: number;
  teamId: number | null;
  fullName: string;
  abbreviation: string;
  position: IPosition | null;
  dateOfBirth: string | null;
  draftYear: number | null;
}

/**
 * API から受け取る試合に紐づく選手情報を表す schema です.
 */
export interface IGamePlayer extends IPlayer {
  jearsyNum: string;
  isHome: boolean;
  isActive: boolean;
  isStarter: boolean;
}
